### PR TITLE
Add Swipe Gesture at TabsController

### LINKED ISSUE #743: TabsController not able to change tab with swipe
The version of Xcode: 8.2.1
The version of iOS: 9+
Material Version: 2.6.3

Recently I am having issue with PageTabBarController. Clicking on tab I am able to change tab. but I am not able to change tabs on swipe.
tabchanges

### PR DESCRIPTION
**Pull request by seubseub**
Fix #743, #823 's swiping between tab content views.

I have added swipe(left, right) gesture at tabs controller view.

It worked well when I tested it on my project, 

but I would appreciate it if you could review it once.

### CODE REVIEW

**Comment by danieldahan** [comment]
I am really excited to review this. Thank you! I will look through it tonight :)

File: Sources/iOS/TabsController.swift [hunk 1]
```diff
@@ -109,6 +109,16 @@ open class TabsController: TransitionController {
     /// The TabBar used to switch between view controllers.
     @IBInspectable
     open let tabBar = TabBar()
+    
+    
+    /// A value that decide viewcontrollers swipeable. default value is false
+    open var isSwipeEnable: Bool = false {
```
Comment (OrkhanAlikhanov) [inline Sources/iOS/TabsController.swift]: I think this should be `isSwipeEnabled` with letter `d` at the end

File: Sources/iOS/TabsController.swift [hunk 2]
```diff
@@ -109,6 +109,16 @@ open class TabsController: TransitionController {
...
+        didSet {
+            if(self.isSwipeEnable) {
```
Comment (OrkhanAlikhanov) [inline Sources/iOS/TabsController.swift]: `if isSwipeEnabled {` would be more clean

File: Sources/iOS/TabsController.swift [hunk 3]
```diff
@@ -340,6 +350,33 @@ extension TabsController {
...
+    }
+    
+    @objc func respondToSwipeGesture(gesture: UIGestureRecognizer) {
+        if let swipeGesture = gesture as? UISwipeGestureRecognizer {
+            switch swipeGesture.direction {
+            case UISwipeGestureRecognizerDirection.right:
+                guard ((self.selectedIndex - 1) >= 0) else {return}
+                self.select(at: self.selectedIndex - 1)
+            case UISwipeGestureRecognizerDirection.left:
+                guard ((self.selectedIndex + 1) < self.viewControllers.count) else {return}
```
Comment (OrkhanAlikhanov) [inline Sources/iOS/TabsController.swift]: What about dropping parentheses and `self.`s ?
```swift
guard (selectedIndex + 1) < viewControllers.count else { return }
```

**Comment by OrkhanAlikhanov** [comment]
Thank you, I needed this, haven't tested yet.

**Comment by OrkhanAlikhanov** [comment]
@seubseub I recommend you following [Swift Style Guide (by raywenderlich.com)](https://github.com/raywenderlich/swift-style-guide)

**Comment by seubseub** [comment]
@OrkhanAlikhanov Thanks! The code has been modified. Thanks for the code review, and a good reference site. I will write this guide in the future :)

File: Sources/iOS/TabsController.swift [hunk 4]
```diff
@@ -340,6 +350,33 @@ extension TabsController {
...
+    
+    /**
+     Prepare Swipe Gesture.
+     */
+    fileprivate func prepareSwipeGesture() {
+        let swipeRight = UISwipeGestureRecognizer(target: self, action: #selector(respondToSwipeGesture(gesture:)))
+        swipeRight.direction = UISwipeGestureRecognizerDirection.right
+        self.view.addGestureRecognizer(swipeRight)
+        let swipeLeft = UISwipeGestureRecognizer(target: self, action: #selector(respondToSwipeGesture(gesture:)))
+        swipeLeft.direction = UISwipeGestureRecognizerDirection.left
```
Comment (OrkhanAlikhanov) [inline Sources/iOS/TabsController.swift]: You can omit `UISwipeGestureRecognizerDirection` actually
```swift
swipeLeft.direction = .left
```

File: Sources/iOS/TabsController.swift [hunk 5]
```diff
@@ -340,6 +350,33 @@ extension TabsController {
...
+        self.view.addGestureRecognizer(swipeLeft)
+    }
+    
+    @objc func respondToSwipeGesture(gesture: UIGestureRecognizer) {
+        if let swipeGesture = gesture as? UISwipeGestureRecognizer {
+            switch swipeGesture.direction {
+            case UISwipeGestureRecognizerDirection.right:
+                guard (selectedIndex - 1) >= 0 else { return }
+                select(at: selectedIndex - 1)
+            case UISwipeGestureRecognizerDirection.left:
```
Comment (OrkhanAlikhanov) [inline Sources/iOS/TabsController.swift]: Same here, `UISwipeGestureRecognizerDirection` can be omitted.

File: Sources/iOS/TabsController.swift [hunk 6]
```diff
@@ -109,6 +109,16 @@ open class TabsController: TransitionController {
     /// The TabBar used to switch between view controllers.
     @IBInspectable
     open let tabBar = TabBar()
+    
+    
+    /// A value that decide viewcontrollers swipeable. default value is false
+    open var isSwipeEnabled: Bool = false {
+        didSet {
+            if isSwipeEnabled {
+                prepareSwipeGesture()
```
Comment (OrkhanAlikhanov) [inline Sources/iOS/TabsController.swift]: `prepareSwipeGesture()` adds gesture recognizers to the `self.view`. And it is called when `isSwipeEnabled` is set to `true`. But it is not removed when `isSwipeEnabled` is disabled. This will lead to issues when a developer decides to change the value `isSwipeEnabled` after he has set it to `true`.

```swift
1. vs.isSwipeEnabled = true // adds gesture recognizers
2. vs.isSwipeEnabled = true // will adds gesture recognizer again
```
```swift
1. vs.isSwipeEnabled = true // adds gesture recognizers
2. vs.isSwipeEnabled = false // will not disable swiping, the `vc` will still be swipeable
```

**Comment by OrkhanAlikhanov** [comment]
@seubseub Sorry, I didn't do my first review thoroughly it was late at night.

File: Sources/iOS/TabsController.swift (hunk 6 above)
Comment (seubseub) [reply Sources/iOS/TabsController.swift]: @OrkhanAlikhanov Yes. This is going to be a problem. I'll fix it and put it back up.

**Comment by seubseub** [comment]
@OrkhanAlikhanov Not at all. Thanks to your review, I was able to better fix the swift style guide. Fixed the previous issue and swift style guide. Thank you for your review. 👍

**Comment by OrkhanAlikhanov** [comment]
@seubseub I just noticed that your branch is 3 commits ahead, 18 commits **behind** `CosmicMind:master`. You'll have to rebase `origin/master` and resolve merge conflicts before the PR becomes mergable, which will add one more commit. I think the PR is becomes dirty. I don't know what @danieldahan thinks in this matter. 


Personally, I would delete all the commits in this PR rebase to `CosmicMind:master` and redo the changes and `git push --force`. If force update does not work I would close this PR and make another clean one-commit PR.

**Comment by seubseub** [comment]
@okankocyigit Thanks to your mention, I change base branch to development latest.

> All pull requests should be made to the development branch. Details should be available describing your fix. Before submitting a pull request, please confirm that merge issues are resolved.

in CONTRIBUTING.md

and I made above 3 commit to one commit. :)

**Comment by adamdahan** [comment]
@seubseub Going to be testing this later in the afternoon with a few other features - will most likely . get merged in Material 3. Will keep you posted. Thank you!

**Comment by adamdahan** [comment]
@seubseub Is it possible to provide a sample project so I can test some edge cases? 

My Gitter handle is: adamdahan

Thank you!

**Comment by seubseub** [comment]
@adamdahan Sure! I will send a sample project to gitter for later testing. 

Thank you!

**Comment by seubseub** [comment]
@adamdahan I sent sample project to your gitter.

**Comment by adamdahan** [comment]
@danieldahan This contribution is good to go! Fully tested and code review was completed by @OrkhanAlikhanov

**Comment by angelix** [comment]
@seubseub 
I think there is a minor bug in this implementation when it comes to delegation.
If swipe is enabled and a user changes a tab with swipe, delegate methods like willSelect, shouldSelect, didSelect are never fired.

**Comment by danieldahan** [comment]
@angelix nice catch. Do you mind opening an issue for this, and we can take a look? Thank you!

**Comment by angelix** [comment]
Sure.

**Comment by seubseub** [comment]
@angelix oh I didn't think at that time.  Thank you for finding bug :)
